Add new guide about overwriting transactions (#1846)
LiskHQ/lisk-docs: docs/modules/ROOT/pages/integrate-blockchain/overwriting-transactions.adoc

= How to overwrite a transaction in the transaction pool
Mona Bärenfänger <mona@lightcurve.io>
:toc: preamble
// URLs
// Project URLs
:url_network_txpool: understand-blockchain/network.adoc#transaction-pool

====
On this page, you'll learn:

* [x] How to overwrite a transaction in the transaction pool
* [x] How to increase the priority of a transaction in the transaction pool
====

== Intro

When a transaction is sent to a node, it is first included in the xref:{url_network_txpool}[transaction pool].
Transactions in the transaction pool wait to be included in a block on the blockchain.

> It may be the case, that the transaction sender intends to update the transaction before it is included in a block.

Reviewer: ```suggestion
imho, I would just slightly reword this short paragraph as follows:
It is possible that the sender of the transaction has the intention of modifying the transaction before it gets included in a block.
```

Author: ```suggestion
It is possible that the sender of the transaction has the intention of modifying/updating the transaction before it gets included in a block.
```

> For example, to update specific parameters of the transaction, to replace the transaction entirely, or to speed up the inclusion of the transaction into a block by providing a higher fee.

Author: ```suggestion
This could involve updating certain parameters of the transaction, completely replacing the transaction, or expediting its inclusion into a block by offering a higher fee.
```

This can be achieved by resending the updated transaction with a higher fee while using the same nonce, as explained in the following steps.

.This only works for transactions in the transaction pool
IMPORTANT: Overwriting a transaction is only possible for transactions which have not yet been included in a block.

When the updated transaction is accepted by the node, it will be added to the transaction pool.

> If the transaction pool already contains the original transaction with the same nonce with the same sender account, validators can choose which transaction to include into the next block.

Reviewer: imho, I would just suggest rewording this sentence as follows:
If the transaction pool already includes the original transaction with the same nonce and sender account, validators have the option to select which transaction to include in the upcoming block.

Author: ```suggestion
f the transaction pool already includes the original transaction with the same nonce and sender account, validators have the option to select which transaction to include in the upcoming block.
```

Generally, a validator will always choose to include the transactions with the highest transaction fees, to maximize their rewards for generating new blocks.

> That is why, if two transactions from the same sender with the same nonce exist in the pool, the transaction with the higher fee will be picked preferably to be included into a block.

Reviewer: ```suggestion
Hence, if two transactions from the same sender with the same nonce exist in the pool, the transaction with the higher fee will be picked preferably to be included into a block.
```

> As soon as the updated transaction is included in a block, it will invalidate the remaining original transaction, because of the outdated nonce.

Reviewer: Once the updated transaction is incorporated into a block, the remaining original transaction will be invalidated due to its outdated nonce.

Author: ```suggestion
Once the updated transaction is incorporated into a block, the remaining original transaction will be invalidated due to its outdated nonce.
```

As a result, the transaction will be removed from the transaction pool.

== Example: Original transaction

> Let's assume we have sent the following transaction to the node, but it is currently pending in the transaction pool, because of too many other transactions waiting in the pool with higher fees:

Reviewer: ```suggestion
Let's assume we have sent the following transaction to the node, but it is currently pending in the transaction pool as there are already too many other transactions waiting in the pool with higher fees:
```

[source,json]
----
{
   "module":"token",
   "command":"transfer",
   "fee":"10000000",
   "nonce":"4",
   "senderPublicKey":"3972849f2ab66376a68671c10a00e8b8b67d880434cc65b04c6ed886dfa91c2c",
   "signatures":[
      "c06cd0ef375911528d491971296532dbd58725f0fccc0f23e6c0656c43e75ecd12635616117e64d005f1c9f2a55de7cf2d8f5cb961927d839d141a18faa80e08"
   ],
   "params":{
      "tokenID":"0400000000000000",
      "amount":"800000000",
      "recipientAddress":"lskt8ovj2shbxrtno8xqqt7cnmzzygdkbt6brnvmj",
      "data":"Happy birthday!"
   },
   "id":"f3f537bbd52464d2f97c02f5ef0f9a805d19ad4f8ef1c7efa1da17cef0e5036a"
}
----

To speed up the inclusion of our transaction into the blockchain, we intend to recreate the exact same transaction with a higher fee.

== Recreating the transaction

Create a new transaction with the same nonce and a higher fee.

A slightly higher fee than specified in the original transaction suffices so that validators will prefer to include the new transaction instead of the old with a lower fee.

> NOTE: In case the network is currently very busy, it can be useful to adjust the fee even higher, so it is included faster into a block than the other transactions in the transaction pool.

Reviewer: ```suggestion
NOTE: If the network is currently experiencing high congestion, increasing the fee even further can prove beneficial, ensuring its faster inclusion into a block compared to other transactions in the transaction pool.
```
